core/src/main/java/com/datastax/oss/driver/internal/core/util/concurrent/PromiseCombiner.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.util.concurrent;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.Promise;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import net.jcip.annotations.ThreadSafe;
+
+/**
+ * A thread-safe version of Netty's {@link io.netty.util.concurrent.PromiseCombiner} that uses
+ * proper synchronization to trigger the completion of the aggregate promise.
+ */
+@ThreadSafe
+public class PromiseCombiner {
+
+  /**
+   * Combines the given futures into the given promise, that is, ties the completion of the latter
+   * to that of the formers.
+   *
+   * @param aggregatePromise The promise that will complete when all parents complete.
+   * @param parents The parent futures.
+   */
+  public static void combine(
```

Comment:
@olim7t I changed your design slightly because it felt odd to instantiate an object without doing anything with it after (that is, all the action was happening in its constructor).